src/VisualStudio/Core/Def/Implementation/Workspace/SolutionSizeTracker.cs

```diff
@@ -0,0 +1,266 @@
+            this._queue = new SimpleTaskQueue(TaskScheduler.Default);
+
+            this._workspace.WorkspaceChanged += OnWorkspaceChanged;
+        }
+
+        public int GetSolutionSize(SolutionId solutionId)
+        {
+            if (_solutionId == solutionId)
+            {
+                return _size;
+            }
+
+            // we don't know solution size
+            return -1;
+        }
+
+        private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
+        {
+            switch (e.Kind)
+            {
+                case WorkspaceChangeKind.SolutionAdded:
+                case WorkspaceChangeKind.SolutionRemoved:
+                case WorkspaceChangeKind.SolutionCleared:
+                    SolutonAddedOrRemoved(e.NewSolution);
+                    return;
+
+                case WorkspaceChangeKind.SolutionReloaded:
+                case WorkspaceChangeKind.SolutionChanged:
+                    SolutionChanged(e.OldSolution, e.NewSolution);
+                    return;
+
+                case WorkspaceChangeKind.ProjectAdded:
+                    ProjectAdded(e.NewSolution.GetProject(e.ProjectId));
+                    return;
+                case WorkspaceChangeKind.ProjectRemoved:
+                    ProjectRemoved(e.OldSolution.GetProject(e.ProjectId));
+                    return;
+
+                case WorkspaceChangeKind.ProjectChanged:
+                case WorkspaceChangeKind.ProjectReloaded:
+                    ProjectChanged(e.OldSolution.GetProject(e.ProjectId), e.NewSolution.GetProject(e.ProjectId));
+                    return;
+
+                case WorkspaceChangeKind.DocumentAdded:
+                    DocumentAdded(e.NewSolution.GetDocument(e.DocumentId));
+                    return;
+
+                case WorkspaceChangeKind.DocumentRemoved:
+                    DocumentRemoved(e.OldSolution.GetDocument(e.DocumentId));
+                    return;
+
+                case WorkspaceChangeKind.DocumentReloaded:
+                case WorkspaceChangeKind.DocumentChanged:
+                    DocumentChanged(e.OldSolution.GetDocument(e.DocumentId), e.NewSolution.GetDocument(e.DocumentId));
+                    return;
+
+                case WorkspaceChangeKind.AdditionalDocumentAdded:
+                case WorkspaceChangeKind.AdditionalDocumentRemoved:
+                case WorkspaceChangeKind.AdditionalDocumentReloaded:
+                case WorkspaceChangeKind.AdditionalDocumentChanged:
+                    return;
+
+                default:
+                    Contract.Requires(false, "Unknown workspace kind");
+                    return;
+            }
+        }
+
+        private void DocumentChanged(Document oldDocument, Document newDocument)
+        {
+            _queue.ScheduleTask(async () =>
+            {
+                var sum = await GetDocumentChangeSizeAsync(oldDocument, newDocument).ConfigureAwait(false);
+                _size = Math.Max(_size + sum, 0);
```

Comment:
changed size can be + or -. ex) user deleted text in a file then delta will be -